src/Main.elm

```diff
@@ -0,0 +1,136 @@
+    }
+
+
+init : Flags -> Url -> Nav.Key -> ( Model, Cmd Msg )
+init flags url navKey =
+    let
+        route =
+            Route.fromUrl flags.basePath url
+
+        preEnv =
+            { flags = flags
+            , route = route
+            , navKey = navKey
+            , perspectiveParams = Route.perspectiveParams route
+            }
+
+        perspectiveToAppInit perspective =
+            let
+                env =
+                    Env.init preEnv.flags perspective
+
+                ( app, cmd ) =
+                    App.init env preEnv.route preEnv.navKey
+            in
+            ( Initialized app, Cmd.map AppMsg cmd )
+
+        fetchPerspective_ =
+            ( Initializing preEnv, Api.perform (ApiBasePath flags.apiBasePath) (fetchPerspective preEnv) )
+    in
+    -- If we have a codebase hash we can construct a full perspective,
+    -- otherwise we have to fetch the hash before being able to start up the
+    -- app
+    preEnv.perspectiveParams
+        |> Perspective.fromParams
+        |> Maybe.map perspectiveToAppInit
+        |> Maybe.withDefault fetchPerspective_
+
+
+fetchPerspective : PreEnv -> ApiRequest Perspective Msg
+fetchPerspective preEnv =
+    Api.list (ByCodebase Relative) "."
+        |> Api.toRequest (Perspective.decode preEnv.perspectiveParams) (FetchPerspectiveFinished preEnv)
+
+
+type Msg
+    = FetchPerspectiveFinished PreEnv (Result Http.Error Perspective)
+    | AppMsg App.Msg
+
+
+update : Msg -> Model -> ( Model, Cmd Msg )
+update msg model =
+    case msg of
+        FetchPerspectiveFinished preEnv result ->
+            case result of
+                Ok perspective ->
+                    let
+                        env =
+                            Env.init preEnv.flags perspective
+
+                        ( app, cmd ) =
+                            App.init env preEnv.route preEnv.navKey
+                    in
+                    ( Initialized app, Cmd.map AppMsg cmd )
+
+                Err err ->
+                    ( InitializationError preEnv err, Cmd.none )
+
+        AppMsg appMsg ->
+            case model of
+                Initialized a ->
+                    let
+                        ( app, cmd ) =
+                            App.update appMsg a
+                    in
+                    ( Initialized app, Cmd.map AppMsg cmd )
+
+                _ ->
+                    ( model, Cmd.none )
+
+
+subscriptions : Model -> Sub Msg
+subscriptions model =
+    case model of
+        Initialized app ->
+            Sub.map AppMsg (App.subscriptions app)
+
+        _ ->
+            Sub.none
+
+
+view : Model -> Browser.Document Msg
+view model =
+    case model of
+        Initializing _ ->
+            { title = "Loading.."
+            , body = [ div [] [ text "Loading.." ] ]
+            }
+
+        InitializationError _ _ ->
+            { title = "Application Error"
+            , body = [ div [] [ text "Application Error" ] ]
+            }
```

Comment:
This is visible at the very edge if the application fails to load completely. We should improve the styling and messaging here. Tracked here #186